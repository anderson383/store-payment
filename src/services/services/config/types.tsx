const TYPES = {PAYMENT_REPOSITORY: Symbol.for('PAYMENT_REPOSITORY'),
  INVENTARY_REPOSITORY: Symbol.for('INVENTARY_REPOSITORY')
};

export {TYPES};
