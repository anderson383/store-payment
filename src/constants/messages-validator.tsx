
export const creditCardRequired = 'El número de tarjeta es obligatorio';

export const creditCardMaxDigit = 'El número de tarjeta debe tener 16 dígitos';

export const cardHolderNameRequired = 'El nombre es obligatorio';

export const expiresRequired = 'La fecha de expiración es obligatoria';

export const cvcRequired = 'El CVC es obligatorio';

export const cvcMaxDigit = 'El CVC debe tener 3 o 4 dígitos';
