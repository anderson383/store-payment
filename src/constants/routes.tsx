import { IndexPage } from '../pages/index';
import { DetailPage } from '../pages/detail/detail';

export const ROUTES:Record<string, {path: string, component: React.ReactNode}> = {
  index: {
    path: '/',
    component: <IndexPage />
  },
  detailproduct: {
    path: '/detail/:id',
    component: <DetailPage />
  }
}