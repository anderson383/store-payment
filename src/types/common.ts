

export type OptionType = {
  label: any;
  value: any;
  meta? : any;
}
