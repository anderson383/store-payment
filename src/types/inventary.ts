export interface ProductType {
  id: string;
  name: string;
  description: string;
  stock: number;
  price: number;
  images: string[];
  size?: number;
  quantity?:number;
}