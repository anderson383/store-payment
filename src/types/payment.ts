export interface CreditCardType {
  cardHolderName: string,
  expires: string
  cardNumber: string
  cvc: string
  numberCuotes: string
}

export interface CustomerInfoType {
  email: string;
  name: string;
  lastname: string;
  phone: string;
  address: string;
  deparment: string;
  city: string;
}