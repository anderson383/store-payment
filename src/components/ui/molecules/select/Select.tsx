import React, {
  useEffect,
  useRef, useState
} from 'react';
import _ from 'lodash';
import styles from './Select.module.scss';
import { useField } from 'formik';
import useClickOutside from '../../../../hooks/use-click-outside';
import { OptionType } from 'types/common';
import ReactDOM from 'react-dom';
import { IconChevronDown } from '@tabler/icons-react';
import { useWindowSize } from '../../../../hooks/use-window';

interface SelectProps {
  name: string;
  label?: string;
  placeholder?: string;
  options: OptionType[];
  viewLabel?: string;
  disabled?: boolean;
  onChange?: (value: OptionType) => void
}

export const Select: React.FC<SelectProps> = ({
  label,
  placeholder,
  name,
  options,
  viewLabel = 'label',
  disabled,
  onChange
}) => {
  const [active, setActive] = useState(false);
  const [field, meta, helpers] = useField(name);
  const portalContainer = document.getElementById('portal-root');
  const selectRef = useRef<HTMLDivElement>(null);
  const { width: widthScreen } = useWindowSize();
  const optionsRef = useRef<any>(null);


  const [popoverPosition, setPopoverPosition] = useState({
    top: 0,
    left: 0,
    width: 0,
    height: 0
  });

  const onReziseScreen = () => {
    if (selectRef.current) {
      const rect = selectRef.current.getBoundingClientRect();

      setPopoverPosition({
        width: rect.width,
        top: rect.bottom,
        left: rect.left,
        height:
          window.innerHeight - rect.bottom > 128 ? null : window.innerHeight - rect.bottom - 16
      });
    }
  };


  useEffect(() => {
    onReziseScreen();
  }, [widthScreen, active]);

  const handleValue = (value: OptionType) => {
    setActive(false);
    helpers.setValue(value);
    onChange && onChange(value);
  };

  useClickOutside(optionsRef, selectRef,() => {
    setActive(false);
  });

  return (
    <div className={styles.select} >
      {label && <label htmlFor={name} >{label}</label>}
      <div className={styles.input_select}  >
        <div onClick={() => !disabled && setActive(!active)} ref={selectRef}>
          <IconChevronDown className={`${ styles.chevron }  ${ active ? styles.chevron_activ : '' }`} />
            <input
              role='textfield'
              tabIndex={0}
              disabled={disabled}
              readOnly
              id={name}
              type="text"
              onBlur={e => {
                helpers.setTouched(true);
                field.onBlur(e);
              }}
              className={styles.selectTrigger}
              placeholder={placeholder}
              value={typeof field.value === 'string' ? field.value : field.value === null ? '' : field.value?.[viewLabel]}
            />
        </div>
          {
            active && ReactDOM.createPortal(
            <div
              ref={optionsRef}
              className={`${ styles.select_options }`}
              style={{
                width: popoverPosition.width,
                top: popoverPosition.top,
                left: popoverPosition.left,
                height: popoverPosition.height,
                pointerEvents: active ? 'auto' : 'auto'
              }}
            >
              {options.map(item => (
                <div
                  className={styles.item_select}
                  key={'select-vaue' + item.value}
                  onClick={() => handleValue(item)}
                >
                  {item.label}
                </div>
              ))}
            </div>,
            portalContainer)
          }
      </div>
      {
       (meta.error && meta.touched) ? <span className={styles.error}> {meta.error} </span> : null
      }
    </div>
  );
};
