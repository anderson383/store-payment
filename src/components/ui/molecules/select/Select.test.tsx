import * as Yup from 'yup';

import {
  fireEvent,
  render,
  screen
} from '@testing-library/react';
import { act } from 'react';
import { Select } from './Select';
import { Formik } from 'formik';
import { fieldRequired } from '../../../../constants/messages-validator';
import { OptionType } from 'types/common';
import { unmountComponentAtNode } from 'react-dom';
import { createRoot } from 'react-dom/client';

const onChangeMock = jest.fn()

jest.mock('')

const validationSchemaText =  Yup.object({
  test: Yup.string().required(fieldRequired)
})

const fieldsOptions: OptionType[]  = [{ label: 'Field 1', value: '1' }, { label: 'Field 2', value: '2' }, { label: 'Field 3', value: '3' } ]


describe('Tests in Select component', () => {
  let container:any = null;
  let root:any = null;

  afterEach(() => {
    act(() => {
      root.unmount();
    })
    container?.remove();
    container = null;
  });

  beforeEach(() => {
    container = document.createElement('div');
    container.setAttribute('id', 'portal-root');
    container.setAttribute('role', 'portal-root');
    root = createRoot(container!);

    document.body.appendChild(container);
  });

  const setup = (ui: React.ReactElement, { initialValues, ...renderOptions }: any) => {
    return render(
      <Formik initialValues={initialValues} onSubmit={() => {}} validationSchema={validationSchemaText} >
        {ui}
      </Formik>,
      renderOptions
    );
  };

  test('Should render the component', async () => {
    const {asFragment} = setup(<Select label='Test' name="test" options={fieldsOptions}/>, { initialValues: { test: null } });
    
    expect(asFragment()).toMatchSnapshot();
  });

  test('selects an option when clicked', async  () => {
    setup(<Select name="select" options={fieldsOptions} onChange={onChangeMock} />, { initialValues: { select: null } });
    const selectInput = screen.getByRole('textfield');
    fireEvent.blur(selectInput); // Click
    await act(() => Promise.resolve())
    fireEvent.click(selectInput); // Abre el menú
    const option = screen.getByText(/Field 1/i);
    fireEvent.click(option); // Selecciona la opción
    expect(onChangeMock).toHaveBeenCalled();
    // expect(selectInput).toHaveValue('Field 1'); // Verifica que el valor sea el correcto
  });

  test('Should open menu and close with click outside', async  () => {
    const { debug, getByRole } = setup(<Select name="select" options={fieldsOptions} onChange={onChangeMock} />, { initialValues: { select: null } });
    const selectInput = screen.getByRole('textfield');
    fireEvent.blur(selectInput); // Click
    await act(() => Promise.resolve())
    fireEvent.click(selectInput); // Abre el menú
    fireEvent.mouseDown(getByRole('portal-root'))
    await act(() => Promise.resolve())
    expect(screen.queryByText(/Field 1/i)).toBeNull();
  });
});
