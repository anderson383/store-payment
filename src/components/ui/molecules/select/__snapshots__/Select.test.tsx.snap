// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in Select component Should render the component 1`] = `
<DocumentFragment>
  <div
    class="select"
  >
    <label
      for="test"
    >
      Test
    </label>
    <div
      class="input_select"
    >
      <div>
        <svg
          class="tabler-icon tabler-icon-chevron-down chevron  "
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6 9l6 6l6 -6"
          />
        </svg>
        <input
          class="selectTrigger"
          id="test"
          readonly=""
          role="textfield"
          tabindex="0"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
